### PR TITLE
feat(ui): Auto-detect URL prefix from window.location at page load

### PR DESCRIPTION
**Pull request by yurishkuro**
Replace the static `<base href="/">` in `index.html` with an inline synchronous script that computes the mount-point prefix from `window.location.pathname` at page-load time, before the browser's preload scanner processes any relative-URL asset references.

**How it works:**
- The script checks `pathname` against every known SPA sub-path (e.g. `/search`, `/trace/`, `/dependencies`), finds the rightmost match, and derives the prefix from everything to its left.
- For unknown routes (e.g. `/pfx/unknown-page`) it falls back to stripping the last path segment, giving the correct parent directory.
- If React fails to mount entirely (assets unreachable, invalid base path), a plain-text message is shown in `#jaeger-ui-root` instead of a blank page.
- Characters that are invalid in URL paths (`& " < >`) are stripped from the computed prefix as a safety measure.

**Why this fixes the reverse-proxy rewrite case (jaeger issue #5157):**

Before this PR, Jaeger requires the external URL prefix to equal the internal URL prefix to equal `--query.base-path`. The proxy cannot rewrite because the backend injects `<base href="/internal/">` into the HTML, and the JS constructs API URLs as `/internal/api/traces`. Those reach the backend correctly only when external == internal.

After this PR, the inline script detects the *external* prefix from `window.location.pathname` (what the browser actually sees) and sets `<base href="/external/">`. The JS then constructs API URLs as `/external/api/traces`. The browser sends those to the proxy, which rewrites `/external/` → `/internal/`, and the backend sees `/internal/api/traces` — exactly the path registered via `--query.base-path=/internal`. This works without any backend change.

The only remaining constraint is that `--query.base-path` must match the **internal** path the proxy forwards to Jaeger — a reasonable and usually static configuration.

**Tests** (`index.test.ts`, next to `index.html`):
- Reads `index.html`, extracts the inline script via the `data-inject-target="BASE_URL"` marker, and executes it via `new Function()` — testing the actual production code rather than a TypeScript mirror.
- Imports `ROUTES` from `App/index.tsx` (the same array used to render `<Routes>`) to structurally enforce that `knownSubPaths` covers every registered route. Adding a route without updating `knownSubPaths` will fail the test automatically.
- 46 test cases covering root prefix, sub-path prefixes, deep prefixes, unknown routes, and rightmost-match edge cases.

See ADR-009 in the jaeger repo for the full design rationale.

Resolves #https://github.com/jaegertracing/jaeger/issues/5157

## AI Usage in this PR (choose one)
- [x] **Heavy**: AI generated most or all of the code changes

### CODE REVIEW

File: packages/jaeger-ui/src/utils/detect-base-path.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2026 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * First-path-segment prefixes of every registered top-level SPA route.
+ * When index.html is served for a deep link (e.g. /jaeger/trace/abc), stripping
+ * the matching sub-path yields the mount-point prefix (/jaeger/).
+ *
+ * Keep this list in sync with the ROUTE_PATH constants in each component's url.ts
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/detect-base-path.ts]: The ADR mentioned that this should be enforced by a unit test, I am not seeing it.

File: packages/jaeger-ui/src/utils/detect-base-path.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+export function detectBasePath(pathname: string): string {
+  for (const sub of KNOWN_SUB_PATHS) {
+    const idx = pathname.indexOf(sub);
+    if (idx !== -1) {
+      const prefix = pathname.slice(0, idx + 1); // up to and including the slash before sub-path
+      return prefix || '/';
+    }
+  }
+  // No known sub-path found: pathname is either "/" or the bare prefix (e.g. "/jaeger/").
+  if (pathname[pathname.length - 1] !== '/') {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/detect-base-path.ts]: can pathname.length be 0?

File: packages/jaeger-ui/index.html [hunk 3]
```diff
@@ -10,9 +10,54 @@
...
+    <!-- NOTE: The document MUST have a <base> element. package.json#homepage is set to "." as part
+         of resolving https://github.com/jaegertracing/jaeger-ui/issues/42 and therefore static
+         assets are linked via relative URLs. This will break on many document URLs, e.g. /trace/abc,
+         unless a valid base URL is provided.
+
+         The base href is set dynamically by the inline script below, which detects the mount-point
+         prefix from window.location at page-load time. This allows a single Jaeger backend to serve
+         the UI correctly under any URL prefix without requiring --query.base-path to match the
+         external prefix, and also works behind URL-rewriting reverse proxies where the external
+         prefix differs from the internal one (see ADR-009). -->
```
Comment (yurishkuro) [inline packages/jaeger-ui/index.html]:
```suggestion
         prefix differs from the internal one (see docs/adr/ADR-009). -->
```

File: packages/jaeger-ui/index.html [hunk 4]
```diff
@@ -10,9 +10,54 @@
...
+    <script>
+      // Detect the URL prefix (mount point) from the current page URL and write a <base href>
+      // element into the document BEFORE the browser's preload scanner encounters any relative-URL
+      // asset references (favicon, scripts). This script must remain the first child of <head>.
+      //
+      // The backend serves index.html for every SPA route under the prefix, so
+      // window.location.pathname may be e.g. "/jaeger/trace/abc123". We strip the known SPA
+      // sub-path suffixes to isolate the mount prefix ("/jaeger/").
```
Comment (yurishkuro) [inline packages/jaeger-ui/index.html]: this comment and the one above are very verbose and will be always included in production index.html, this is overkill. We can simply say
```
<!-- The document MUST have a <base> element, see jaeger/docs/adr/ADR-009 -->
```

File: packages/jaeger-ui/index.html [hunk 5]
```diff
@@ -10,9 +10,40 @@
...
     <meta http-equiv="pragma" content="no-cache" />
 
-    <!-- NOTE: The document MUST have a <base> element. package.json#homepage is set to "." as part of resolving https://github.com/jaegertracing/jaeger-ui/issues/42 and therefore static assets are linked via relative URLs. This will break on many document URLs, e.g. /trace/abc, unless a valid base URL is provided. The base href defaults to "/" but the query-service can inject an override. -->
-    <base href="/" data-inject-target="BASE_URL" />
-    <link rel="shortcut icon" href="/favicon.ico" />
+    <!-- The document MUST have a <base> element, see jaeger/docs/adr/ADR-009 -->
+    <script>
+      // Sets <base href> from window.location BEFORE the preload scanner sees any relative-URL
+      // asset tags. Must remain the first child of <head>. Keep knownSubPaths in sync with
+      // detect-base-path.ts and each component's url.ts (enforced by unit tests).
```
Comment (yurishkuro) [inline packages/jaeger-ui/index.html]: remove comment

File: packages/jaeger-ui/index.html [hunk 6]
```diff
@@ -10,9 +10,54 @@
...
+            break;
+          }
+        }
+        // Ensure trailing slash — required for <base href> to resolve relative URLs correctly.
+        if (prefix[prefix.length - 1] !== '/') {
+          prefix += '/';
+        }
+        document.currentScript.insertAdjacentHTML(
+          'afterend',
+          '<base href="' + prefix + '" data-inject-target="BASE_URL" />'
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.html]: Fixed in commit bd2a9967: added HTML escaping for the prefix before insertAdjacentHTML.

File: packages/jaeger-ui/src/utils/detect-base-path.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ *   "/jaeger/trace/abc123"  → "/jaeger/"
+ */
+export function detectBasePath(pathname: string): string {
+  for (const sub of KNOWN_SUB_PATHS) {
+    const idx = pathname.indexOf(sub);
+    if (idx !== -1) {
+      const prefix = pathname.slice(0, idx + 1); // up to and including the slash before sub-path
+      return prefix || '/';
+    }
+  }
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/detect-base-path.ts]: Fixed: detect-base-path.ts has been deleted and replaced by an inline script in index.html that uses lastIndexOf across all knownSubPaths to find the rightmost match. The test in index.test.ts covers this case explicitly.

File: packages/jaeger-ui/src/utils/detect-base-path.test.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    });
+
+    it('strips /a/b/c/search to /a/b/c/', () => {
+      expect(detectBasePath('/a/b/c/search')).toBe('/a/b/c/');
+    });
+
+    it('strips /a/b/c/trace/deadbeef to /a/b/c/', () => {
+      expect(detectBasePath('/a/b/c/trace/deadbeef')).toBe('/a/b/c/');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/detect-base-path.test.ts]: Fixed: detect-base-path.test.ts has been deleted. The test coverage including rightmost-match edge cases now lives in index.test.ts alongside index.html.

File: packages/jaeger-ui/src/utils/detect-base-path.ts (hunk 1 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/detect-base-path.ts]: Fixed: detect-base-path.ts deleted. Enforcement is now done via index.test.ts which imports ROUTES from App/index.tsx (the same array used to render the actual routes), making gaps structurally impossible.

File: packages/jaeger-ui/src/utils/detect-base-path.ts (hunk 2 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/detect-base-path.ts]: Fixed: the script now uses window.location.pathname || '/' so an empty pathname falls back to '/' instead of causing issues.

File: packages/jaeger-ui/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+// ROUTES is the array rendered directly into <Routes> in App/index.tsx.
+// Importing it here means the coverage test automatically covers any new
+// route added to the router without requiring manual updates to this file.
+import { ROUTES } from './src/components/App';
+
+// Extract and compile the base-path detection script from index.html once
+// to test the base path determination logic.
+const indexHtml = readFileSync(join(dirname(fileURLToPath(import.meta.url)), 'index.html'), 'utf-8');
+const scriptMatch = indexHtml.match(/<script>([\s\S]*?)<\/script>/);
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.test.ts]: Fixed in 4e47f087: replaced the bare <script> regex with a marker-based approach. The inline script now starts with /* BASE_PATH_DETECT */ and the test matches on that marker, which is robust to tag attributes and script reordering.

File: packages/jaeger-ui/src/components/App/index.tsx [hunk 10]
```diff
@@ -42,34 +42,43 @@ processScripts();
 // to tree-shake the dynamic import and exclude the demo files from the prod bundle.
 const PlexusDemoPage = import.meta.env.DEV ? React.lazy(() => import('../PlexusDemo')) : null;
 
+// The route table is exported so that detect-base-path.test.ts can verify that
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/components/App/index.tsx]: Fixed in 4e47f087: updated the comment to reference index.test.ts.

File: packages/jaeger-ui/index.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.html]: The reference to detect-base-path.ts was removed in an earlier commit. The current index.html has a short comment pointing to ADR-009 and the BASE_PATH_DETECT marker for the test.

File: packages/jaeger-ui/index.html [hunk 11]
```diff
@@ -10,9 +10,40 @@
...
+      // asset tags. Must remain the first child of <head>. Keep knownSubPaths in sync with
+      // detect-base-path.ts and each component's url.ts (enforced by unit tests).
+      (function () {
+        var knownSubPaths = [
+          '/deep-dependencies',
+          '/dependencies',
+          '/quality-metrics',
+          '/plexus-demo',
+          '/monitor',
+          '/search',
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.html:22]: Fixed in 4e47f087: added '/trace' (without trailing slash) to knownSubPaths alongside '/trace/', so bare /trace URLs are handled correctly.

File: packages/jaeger-ui/index.test.ts [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+if (!scriptMatch) throw new Error('No inline script found in index.html');
+// eslint-disable-next-line no-new-func
+const scriptFn = new Function('document', 'window', scriptMatch[1]);
+
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.test.ts:24]: The approach of testing the actual inline script from index.html via new Function() is intentional — it tests real production code rather than a TypeScript mirror. The PR description has been updated accordingly.

File: packages/jaeger-ui/index.test.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.test.ts]: Fixed in 4e47f087: replaced the bare <script> regex with a marker comment approach. The script now starts with /* BASE_PATH_DETECT */ and the test extracts the script using that marker.

File: packages/jaeger-ui/index.html [hunk 13]
```diff
@@ -10,26 +10,57 @@
...
+          '/trace/',
+          '/trace',
+        ];
+        var path = window.location.pathname || '/';
+        var bestIdx = -1;
+        for (var i = 0; i < knownSubPaths.length; i++) {
+          var idx = path.lastIndexOf(knownSubPaths[i]);
+          if (idx !== -1 && idx > bestIdx) bestIdx = idx;
+        }
+        var prefix = bestIdx !== -1 ? path.slice(0, bestIdx + 1) || '/' : path;
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.html]: This is an inherent ambiguity in the URL: /a/search/ is correctly interpreted as 'prefix=/a/, route=/search' — that is exactly the case a user would encounter when their Jaeger is mounted at /a/ and they navigate to the search page. There is no content-only URL that unambiguously means 'prefix is /a/search/, route is /' because Jaeger doesn't serve index.html for bare prefix-only URLs in practice. The script's behavior here is correct.

File: packages/jaeger-ui/src/components/App/index.tsx [hunk 14]
```diff
@@ -42,34 +42,43 @@ processScripts();
 // to tree-shake the dynamic import and exclude the demo files from the prod bundle.
 const PlexusDemoPage = import.meta.env.DEV ? React.lazy(() => import('../PlexusDemo')) : null;
 
+// The route table is exported so that index.test.ts can verify that
+// KNOWN_SUB_PATHS covers every path registered here. The <Routes> below is rendered
+// directly from this array, so the two cannot diverge.
+export const ROUTES: { path: string; element: React.ReactNode }[] = [
+  { path: searchPath, element: <SearchTracePage /> },
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/components/App/index.tsx:49]: The side-effect concern is theoretical. The tests run cleanly with no warnings or errors (verified). processScripts() is a no-op in jsdom (no external scripts to process), and JaegerAPI.apiRoot assignment is a harmless field set. Refactoring ROUTES into a separate module would add complexity without fixing an actual problem.

File: packages/jaeger-ui/index.test.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+import { readFileSync } from 'fs';
+import { dirname, join } from 'path';
+import { fileURLToPath } from 'url';
+
+// ROUTES is the array rendered directly into <Routes> in App/index.tsx.
+// Importing it here means the coverage test automatically covers any new
+// route added to the router without requiring manual updates to this file.
+import { ROUTES } from './src/components/App';
+
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.test.ts:12]: Same response as the comment on App/index.tsx: the side effects are harmless in the test environment. Tests pass cleanly with no warnings. The import is intentional — it ensures the enforcement test automatically covers any new route added to the router.

File: packages/jaeger-ui/index.html [hunk 16]
```diff
@@ -10,26 +10,61 @@
...
+        var path = window.location.pathname || '/';
+        var bestIdx = -1;
+        for (var i = 0; i < knownSubPaths.length; i++) {
+          var idx = path.lastIndexOf(knownSubPaths[i]);
+          if (idx !== -1 && idx > bestIdx) bestIdx = idx;
+        }
+        var prefix =
+          bestIdx !== -1
+            ? path.slice(0, bestIdx + 1) || '/'
+            : path.slice(0, path.lastIndexOf('/') + 1) || '/';
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/index.html:35]: Invalid: Jaeger's ServeMux registers the catch-all at basePath+'/' (with trailing slash). A bare /jaeger (no slash) never matches that pattern and returns 404 — index.html is never served for it, so the script never runs. In deployments with a reverse proxy, a 301 redirect /jaeger → /jaeger/ is configured (as in our httpd.conf), so the browser always arrives at the trailing-slash form before index.html is served.
